beacon-chain/db/kv/state.go

```diff
@@ -264,3 +298,73 @@ func createState(enc []byte) (*pb.BeaconState, error) {
 	}
 	return protoState, nil
 }
+
+// HighestSlotState returns the state with the highest slot from the db.
+func (k *Store) HighestSlotState(ctx context.Context) (*state.BeaconState, error) {
+	ctx, span := trace.StartSpan(ctx, "BeaconDB.HighestSlotState")
+	defer span.End()
+	var s *pb.BeaconState
+	err := k.db.View(func(tx *bolt.Tx) error {
+		slotBkt := tx.Bucket(slotsHasObjectBucket)
+		savedSlots := slotBkt.Get(savedStateSlotsKey)
+		highestIndex, err := bytesutil.HighestBitIndex(savedSlots)
+		if err != nil {
+			return err
+		}
+		highestSlot := highestIndex - 1
+		f := filters.NewFilter().SetStartSlot(uint64(highestSlot)).SetEndSlot(uint64(highestSlot))
+
+		keys, err := getBlockRootsByFilter(ctx, tx, f)
+		if err != nil {
+			return err
+		}
+		if len(keys) != 1 {
```

Comment:
Why this constraint? Sometimes there are multiple block roots at a given slot.

Reply:
Multiple block roots at a given slot means a double propose and it could happen. I updated the API to:
```
HighestSlotStates(ctx context.Context) ([]*state.BeaconState, error)
```